operator/buffer/circular_file.go

```diff
@@ -0,0 +1,327 @@
+package buffer
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"go.uber.org/multierr"
+)
+
+// circularFile is a io.ReadWriteCloser that writes to a fixed length file, such that
+// it wraps around to the beginning when reaching the end.
+// Methods on this struct are not thread-safe and require additional synchronization.
+type circularFile struct {
+	Start      int64
+	ReadPtr    int64
+	End        int64
+	Size       int64
+	Full       bool
+	f          fileLike
+	seekedRead bool
+	seekedEnd  bool
+	closed     bool
+	readPtrEnd bool
+}
+
+var _ io.ReadWriteCloser = (*circularFile)(nil)
+
+func openCircularFile(filePath string, sync bool, size int64) (*circularFile, error) {
+	fileFlags := os.O_CREATE | os.O_RDWR
+	if sync {
+		fileFlags |= os.O_SYNC
+	}
+
+	f, err := os.OpenFile(filePath, fileFlags, 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make sure that the file, if it already existed, is actually the correct size.
+	fsize, err := f.Seek(0, io.SeekEnd)
+	if err != nil {
+		fCloseErr := f.Close()
+		return nil, multierr.Combine(
+			err,
+			fCloseErr,
+		)
+	}
+
+	if fsize == 0 {
+		err := f.Truncate(size)
+		if err != nil {
+			fCloseErr := f.Close()
+			return nil, multierr.Combine(
+				err,
+				fCloseErr,
+			)
+		}
+	} else if fsize != size {
```

Comment:
Seems it would be ok if `size > fsize` because we can grow but not if `size < fsize`. Is this true or do we need to worry if the size does not match?

Reply:
The size needs to match exactly (we cannot grow), because the wrap around point would change. So if you have an entry that spans across the boundary, growing would completely break it.